controllers/topology/reconcile_state.go

```diff
@@ -16,10 +16,187 @@ limitations under the License.
 
 package topology
 
-import "context"
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
 
 // reconcileState reconciles the current and desired state of the managed Cluster topology.
+// NOTE: We are assuming all the required objects are provided as input; also, in case of any error,
+// the entire reconcile operation will fail. This might be improved in the future if support for reconciling
+// subset of a topology will be implemented.
 func (r *ClusterReconciler) reconcileState(ctx context.Context, current, desired *clusterTopologyState) error {
-	// TODO: add reconcile logic
+	// Reconcile desired state of the InfrastructureCluster object.
+	if err := r.reconcileInfrastructureCluster(ctx, current, desired); err != nil {
+		return err
+	}
+
+	// TODO: reconcile control plane
+
+	// Reconcile desired state of the InfrastructureCluster object.
+	if err := r.reconcileCluster(ctx, current, desired); err != nil {
+		return err
+	}
+
+	// TODO: reconcile machine deployments
+
+	return nil
+}
+
+// reconcileInfrastructureCluster reconciles the desired state of the Cluster object.
+// NOTE: this assumes reconcileInfrastructureCluster and reconcileControlPlane being already completed;
+// most specifically, after a cluster is created it is assumed that the reference to the infrastructure cluster /
+// control plane should never change (only the content of the objects can eventually change).
+func (r *ClusterReconciler) reconcileCluster(ctx context.Context, current, desired *clusterTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Check differences between current and desired state, and eventually patch the current object.
+	patchHelper, err := newMergePatchHelper(current.cluster, desired.cluster, r.Client)
+	if err != nil {
+		return errors.Wrap(err, "failed to create patch helper for the Cluster object")
+	}
+	if patchHelper.HasChanges() {
+		log.Info("updating Cluster")
```

Comment:
I will defer this to https://github.com/kubernetes-sigs/cluster-api/issues/5067, but for this specific case the answer is no because the controller runtime logger already adds all the info about the cluster